RichTextLabel Highlighting, Redacting

golfinq opened the pull request:
This pull request addresses godotengine/godot-proposals#367 by adding highlighting and redacting capabilities to RichTextLabel with the tags bgcolor= and fgcolor=

Edit: Removed outdated information

golfinq committed:
Rich Text Label now allows for foreground colors and background colors

Chaosus commented:
Don't forget to merge the commits together as it requested by our workflow, see http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase

dalexeev commented on scene/gui/rich_text_label.cpp:
```diff
@@ -2752,6 +2805,8 @@ void RichTextLabel::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_effects"), &RichTextLabel::get_effects);
 	ClassDB::bind_method(D_METHOD("install_effect", "effect"), &RichTextLabel::install_effect);
 
+	ClassDB::bind_method(D_METHOD("parse_color"), &RichTextLabel::parse_color);
```
Please note that in 4.0 it is planned to switch to standard named colors from the `Color` class (#35505).

golfinq replied:
I just pushed a commit which utilizes the changes made to the `Color` class in #35505

dalexeev commented:
Also, the code for parsing BB tags is duplicated in the following file. :disappointed: 

https://github.com/godotengine/godot/blob/5db45fbaf351883e37fdcdece854139b6b236d4b/editor/editor_help.cpp#L1391-L1400

Calinou commented:
What's the difference between `[fgcolor]` and `[color]`?

golfinq commented:
Please note for commit to work for 3.2; change the parse_color function to below. As I am assuming #35505 compatibility in the commit.
 
`Color RichTextLabel::parse_color(const String &p_color) {
	Color color;

	if (Color::html_is_valid(col))
		color = Color::html(col);
	else
		color = Color::named(col);

	if (color == Color() && color != "black" && color != "#000000")
		color = get_color("default_color");

	return color;
}`

dalexeev commented on scene/gui/rich_text_label.cpp:
```diff
@@ -2907,6 +2964,19 @@ Dictionary RichTextLabel::parse_expressions_for_values(Vector<String> p_expressi
+Color RichTextLabel::parse_color(const String &p_color) {
+	Color color;
+
+	if (Color::html_is_valid(col))
```
The variable `col` is not declared. Instead, there should be a `p_color`. :smile:

golfinq replied:
Thank you for checking :smiley:

dalexeev commented:
> ```
> 	if (color == Color() && color != "black" && color != "#000000")
> ```

In the second and third conditions, instead of `color`, should there be `p_color`? :wink:

golfinq commented:
I updated the code to be more modular, I also changed the selection behavior of the `fgcolor` tag to change the color of the selected part to the selection color.

dalexeev commented:
@golfinq #35505 was finally merged, so you can rebase your PR. 😃

golfinq commented:
This pull request is now updated and It _should_ work with all of the characters `RichTextLabel` is now expected to draw

Calinou reviewed (approved):
Feature looks good to me. We'll need it at some point to improve the display of the built-in editor help (e.g. making inline code stand out more).

Make sure to [update the class reference XML](https://docs.godotengine.org/en/latest/community/contributing/updating_the_class_reference.html) to document the newly added methods.

golfinq commented:
I am unsure what I did, but I just wanted to request review, not remove one. Sorry.

KoBeWi commented:
The background doesn't properly fit to letters:
![image](https://user-images.githubusercontent.com/2223172/122475502-0ab3e200-cfc5-11eb-8f23-e62a8b0d159c.png)
Compare with e.g. Google Docs:
![image](https://user-images.githubusercontent.com/2223172/122475554-1dc6b200-cfc5-11eb-96d1-eb79bda4a46d.png)

Also this doesn't fully resolve the proposal, as the user wanted the "redacted" text to be unselectable. This is fine and doesn't need to be implemented here, but the OP needs to be updated not to auto-close the proposal.

Calinou commented:
> Also this doesn't fully resolve the proposal, as the user wanted the "redacted" text to be unselectable. This is fine and doesn't need to be implemented here, but the OP needs to be updated not to auto-close the proposal.

To draw redacted text that cannot be retrieved by external means, it's likely better to replace the text with a Unicode character (and keep spaces as-is if desired).

golfinq commented:
> The background doesn't properly fit to letters

I figured out a solution and now properly fills the entire line.
![Image](https://i.ibb.co/FmBCyst/godot-example-1-editor.png)

> the user wanted the "redacted" text to be unselectable.

To prevent copy/paste will be more tricky and would probably be best done in a separate pull request as you mentioned. My implementation would be to create a flag in each `Glyph` object which would tell the text server whether or not to draw the actual characters, but keep track of the spacing for everything else. Then find where Godot communicates with the host OS to fill the clipboard and skip any character which has that flag set.

Calinou commented:
> Would this work with fonts that do not support unicode?

Yes, by defining a fallback font that contains the character you need. It will be used if the main font doesn't contain the glyph in question.

bruvzg reviewed (changes requested):
There's an issue with this approach (drawing highlights for each glyph individually) if you are using font/script with the overlapping glyphs.

<img width="263" alt="Screenshot 2021-06-19 at 10 49 14" src="https://user-images.githubusercontent.com/7645683/122635450-adb34b80-d0ec-11eb-9d2b-47f0b708a9bb.png">


On the screenshot, semi transparent `bgcolor` is covering part of the text.

For this reason, text selection is fully drawn before drawing any text(see lines 940-953), same should be done for highlights:

1. Draw selection and `bgcolor` rectangles.
2. Draw text outlines.
3. Draw main text.
4. Draw `fgcolor` rectangles.

Text server have dedicated for this specific purpose, which return non overlapping highlight/selection ranges for the portion of text:

```
Vector<Vector2> range = TS->shaped_text_get_selection(rid, start_position, end_position);
```

golfinq commented:
I believe it is now working as expected. Here is my test text

```
[fgcolor=red]afdasagrffag[/fgcolor]asdf
Lorem ipsum dolor sit amet,
 [bgcolor=#0000FF22] الحدود وصافرات, مع و [/bgcolor] aing elit, sed
[bgcolor=#00000044]do eiu smod[bgcolor=black] वंबर जैसी अत्यंत क्ष[/bgcolor] incididunt[/bgcolor] 
rr
ff [bgcolor=red]将エ管話ツナfffgg ルヒ社[/bgcolor] ff

test [bgcolor=black]adsfg[bgcolor=yellow]afg[/bgcolor]fdasfdasfj[/bgcolor]
```
Here is the output
![image](https://i.ibb.co/WBwKb2H/godot-example-2-editor.png)

bruvzg commented on scene/gui/rich_text_label.h:
```diff
@@ -437,6 +451,8 @@ class RichTextLabel : public Control {
 	Ref<RichTextEffect> _get_custom_effect_by_code(String p_bbcode_identifier);
 	virtual Dictionary parse_expressions_for_values(Vector<String> p_expressions);
 
+	void _draw_fbg_boxes(RID p_ci, RID p_rid, Vector2 line_off, Item *it_from, Item *it_to, const TextServer::Glyph *glyphs, int gl_size, int fbg_flag);
```
```suggestion
	void _draw_fbg_boxes(RID p_ci, RID p_rid, Vector2 line_off, Item *it_from, Item *it_to, int start, int end, int fbg_flag);
```

bruvzg commented on scene/gui/rich_text_label.cpp:
```diff
@@ -934,6 +930,10 @@ int RichTextLabel::_draw_line(ItemFrame *p_frame, int p_line, const Vector2 &p_o
 
+		Vector2 fbg_line_off = off + p_ofs;
+		// Draw background color box
+		_draw_fbg_boxes(ci, rid, fbg_line_off, it_from, it_to, glyphs, gl_size, 0);
```
```suggestion
		Vector2i chr_range = TS->shaped_text_get_range(rid);
		_draw_fbg_boxes(ci, rid, fbg_line_off, it_from, it_to, chr_range.x, chr_range.y, 0);
```

bruvzg commented on scene/gui/rich_text_label.cpp:
```diff
@@ -1079,6 +1082,9 @@ int RichTextLabel::_draw_line(ItemFrame *p_frame, int p_line, const Vector2 &p_o
 			}
 		}
+		// Draw foreground color box
+		_draw_fbg_boxes(ci, rid, fbg_line_off, it_from, it_to, glyphs, gl_size, 1);
```
```suggestion
		_draw_fbg_boxes(ci, rid, fbg_line_off, it_from, it_to, chr_range.x, chr_range.y, 1);
```

bruvzg commented on scene/gui/rich_text_label.cpp:
```diff
@@ -4108,6 +4178,66 @@ Size2 RichTextLabel::get_minimum_size() const {
 	return size;
 }
 
+void RichTextLabel::_draw_fbg_boxes(RID p_ci, RID p_rid, Vector2 line_off, Item *it_from, Item *it_to, const TextServer::Glyph *glyphs, int gl_size, int fbg_flag) {
```
```suggestion
void RichTextLabel::_draw_fbg_boxes(RID p_ci, RID p_rid, Vector2 line_off, Item *it_from, Item *it_to, int start, int end, int fbg_flag) {
```

bruvzg commented on scene/gui/rich_text_label.cpp:
```diff
@@ -4108,6 +4178,66 @@ Size2 RichTextLabel::get_minimum_size() const {
+	Vector2i fbg_index = Vector2i(INT_MAX, INT_MIN);
+	Color last_color = Color(0, 0, 0, 0);
+	bool draw_box = false;
+	// Draw a box based on color tags associated with glyphs
+	for (int i = 0; i < gl_size; i++) {
+		Item *it = _get_item_at_pos(it_from, it_to, glyphs[i].start);
+		Color color = Color(0, 0, 0, 0);
```
```suggestion
	// Draw a box based on color tags associated with glyphs
	for (int i = start; i < end; i++) {
		Item *it = _get_item_at_pos(it_from, it_to, i);
		Color color = Color(0, 0, 0, 0);
```

bruvzg commented on scene/gui/rich_text_label.cpp:
```diff
@@ -4108,6 +4178,66 @@ Size2 RichTextLabel::get_minimum_size() const {
+		}
+
+		if (draw_box) {
+			Vector<Vector2> sel = TS->shaped_text_get_selection(p_rid, glyphs[fbg_index.x].start, glyphs[fbg_index.y].start);
```
```suggestion
			Vector<Vector2> sel = TS->shaped_text_get_selection(p_rid, fbg_index.x, fbg_index.y);
```

bruvzg commented on scene/gui/rich_text_label.cpp:
```diff
@@ -4108,6 +4178,66 @@ Size2 RichTextLabel::get_minimum_size() const {
+	}
+
+	if (last_color.a > 0) {
+		fbg_index.y = MAX(gl_size - 1, fbg_index.y);
+		Vector<Vector2> sel = TS->shaped_text_get_selection(p_rid, glyphs[fbg_index.x].start, glyphs[fbg_index.y].end);
```
```suggestion
		Vector<Vector2> sel = TS->shaped_text_get_selection(p_rid, fbg_index.x, end);
```

bruvzg reviewed (changes requested):
Using over glyph array won't work if there are writing direction changes in the middle of highlight, using character range instead should work.

golfinq commented:
> Using over glyph array won't work if there are writing direction changes in the middle of highlight, using character range instead should work.

It worked! For reference here is my test text where I believe exactly what you describe happens:
```
ipsum dolor
[fgcolor=red]afdasagrffag[/fgcolor]
Lorem ipsum dolor sit amet,
[bgcolor=#0000FF22] الحدود وصافرات, مع وkkl ghjt[/bgcolor] aing elit, sed
[bgcolor=#00000044]do eiu smod[bgcolor=black] वंबर जैसी अत्यंत क्ष[/bgcolor] incididunt[/bgcolor][fgcolor=black]quickjjkkgG[/fgcolor]
     [bgcolor=black]across lines
rr[/bgcolor]jk
ff [bgcolor=red]将エ管話ツナfffgg ルヒ社[/bgcolor] ff

test [bgcolor=black]adsfg[bgcolor=yellow]afg[/bgcolor]fdasfdasfj[/bgcolor]
```

and here is the result

![image](https://i.ibb.co/ZxH0hkP/godot-example-3-editor.png)

bruvzg commented on scene/gui/rich_text_label.cpp:
```diff
@@ -4108,6 +4183,65 @@ Size2 RichTextLabel::get_minimum_size() const {
+
+		if (draw_box) {
+			Vector<Vector2> sel = TS->shaped_text_get_selection(p_rid, fbg_index.x, fbg_index.y);
+			for (int i = 0; i < sel.size(); i++) {
+				Vector2 rect_off = line_off + Vector2(sel[i].x, -TS->shaped_text_get_ascent(p_rid));
+				Vector2 rect_size = Vector2(sel[i].y - sel[i].x, TS->shaped_text_get_size(p_rid).y);
+				RenderingServer::get_singleton()->canvas_item_add_rect(p_ci, Rect2(rect_off, rect_size), last_color);
+			}
```
```suggestion
			for (int j = 0; j < sel.size(); j++) {
				Vector2 rect_off = line_off + Vector2(sel[j].x, -TS->shaped_text_get_ascent(p_rid));
				Vector2 rect_size = Vector2(sel[j].y - sel[j].x, TS->shaped_text_get_size(p_rid).y);
				RenderingServer::get_singleton()->canvas_item_add_rect(p_ci, Rect2(rect_off, rect_size), last_color);
			}
```

bruvzg commented on scene/gui/rich_text_label.cpp:
```diff
@@ -4108,6 +4183,65 @@ Size2 RichTextLabel::get_minimum_size() const {
+				Vector2 rect_size = Vector2(sel[i].y - sel[i].x, TS->shaped_text_get_size(p_rid).y);
+				RenderingServer::get_singleton()->canvas_item_add_rect(p_ci, Rect2(rect_off, rect_size), last_color);
+			}
+			fbg_index = Vector2i(INT_MAX, INT_MIN);
```
```suggestion
			fbg_index = Vector2i(end, start);
```

bruvzg commented on scene/gui/rich_text_label.cpp:
```diff
@@ -4108,6 +4183,65 @@ Size2 RichTextLabel::get_minimum_size() const {
 }
 
+void RichTextLabel::_draw_fbg_boxes(RID p_ci, RID p_rid, Vector2 line_off, Item *it_from, Item *it_to, int start, int end, int fbg_flag) {
+	Vector2i fbg_index = Vector2i(INT_MAX, INT_MIN);
```
```suggestion
	Vector2i fbg_index = Vector2i(end, start);
```

bruvzg reviewed (commented):
CI log:
```
scene/gui/rich_text_label.cpp:4187:32: error: 'INT_MAX' was not declared in this scope
 4187 |  Vector2i fbg_index = Vector2i(INT_MAX, INT_MIN);
      |                                ^~~~~~~
scene/gui/rich_text_label.cpp:46:1: note: 'INT_MAX' is defined in header '<climits>'; did you forget to '#include <climits>'?
   45 | #include "editor/editor_scale.h"
  +++ |+#include <climits>
   46 | #endif
scene/gui/rich_text_label.cpp:4187:41: error: 'INT_MIN' was not declared in this scope
 4187 |  Vector2i fbg_index = Vector2i(INT_MAX, INT_MIN);
      |                                         ^~~~~~~
scene/gui/rich_text_label.cpp:4187:41: note: 'INT_MIN' is defined in header '<climits>'; did you forget to '#include <climits>'?
scene/gui/rich_text_label.cpp:4218:13: error: declaration of 'i' shadows a previous local [-Werror=shadow=local]
 4218 |    for (int i = 0; i < sel.size(); i++) {
      |             ^
scene/gui/rich_text_label.cpp:4191:11: note: shadowed declaration is here
 4191 |  for (int i = start; i < end; i++) {
      |           ^
```

golfinq commented:
I believe the code should pass all checks, but if not. I'll keep plugging away until it does.

akien-mga commented on scene/gui/rich_text_label.cpp:
```diff
@@ -4056,6 +4129,8 @@ void RichTextLabel::_bind_methods() {
 	BIND_ENUM_CONSTANT(ITEM_WAVE);
 	BIND_ENUM_CONSTANT(ITEM_TORNADO);
 	BIND_ENUM_CONSTANT(ITEM_RAINBOW);
+	BIND_ENUM_CONSTANT(ITEM_FGCOLOR);
+	BIND_ENUM_CONSTANT(ITEM_BGCOLOR);
```
Those two should be swapped to match the enum order, otherwise the enum values are wrongly ordered in the docs.

golfinq replied:
I swapped them in both `doc/classes/RichTextLabel.xml` and `scene/gui/rich_text_label.cpp` , it now matches the order in the `enum` list

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:

PseudoC0de commented:
Any chance to have this back-ported to **3.x**? It would be very useful to have this available while **4.x** releases and catches up in stability/production-usability.

golfinq commented:
When this was first written it was in 3.x, but was creating the box per glyph; which lead to issues with international alphabets and wasn't great at identifying the proper offsets for the line it was filling. Subsequently, the PR was entirely rewritten using the `master` branch glyph position handling to identify the height and where to start and stop drawing the boxes. It has been a while since I've looked at the 3.x source code; but it might be possible to borrow the underline positioning algorithm to backport to 3.x as the tag stack filling and identifying mostly stayed the same. Unfortunately, I am not in a position to volunteer code at this time, but that would be the path I would pursue.
